thriftbp/ttl_client.go

```diff
@@ -97,15 +114,59 @@ func (c *ttlClient) IsOpen() bool {
 
 // refresh is called when the ttl hits to try to refresh the connection.
 func (c *ttlClient) refresh() {
-	client, transport, err := c.generator()
-	if err != nil {
+	// set up ctx for retry cancellation
+	var ctx context.Context
+	var closeCalled bool
+	{
+		state := <-c.state
+		ctx, state.cancelRefresh = context.WithCancel(context.Background())
+		defer state.cancelRefresh()
+		closeCalled = state.closed
+		c.state <- state
+	}
+
+	if closeCalled {
+		// In a rare race condition, it's possible that the timer for refresh fired
+		// at the time Close is called. In such case, return early here to avoid
+		// getting into a very long retry loop later with no way to cancel.
+		return
+	}
+	var client thrift.TClient
+	var transport thrift.TTransport
+	var attempts int
+	defer func() {
+		ttlClientRefreshAttemptsHisto.With(prometheus.Labels{
+			clientNameLabel: c.slug,
+		}).Observe(float64(attempts))
+	}()
+	if retrybp.Do(
+		ctx,
+		func() error {
+			attempts++
+			var err error
+			client, transport, err = c.generator()
+			if err != nil {
+				ttlClientReplaceCounter.With(prometheus.Labels{
+					clientNameLabel: c.slug,
+					successLabel:    prometheusbp.BoolString(false),
+				}).Inc()
+			}
+			return err
+		},
+		retry.Attempts(ttlClientRefreshAttempts),
```

Comment:
I'm not familiar with `retrybp`, but just confirming it is intentional to use this `retry.Attempts` from the avast package? If so might be worth replicating it in retrybp, but that's sort of separate from this PR

Reply:
retrybp is the baseplate wrapper on retry, we don't need to replicate everything when we are not really adding anything.